File: Python/Medium/1930.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        alphabet= "abcdefghijklmnopqrstuvwxyz"
        count = 0
        for c in alphabet:
            left = s.find(c)   # return the smallest index that contains c
            right = s.rfind(c) # return the largest index that contains c

            # if c is not found then returns -1 also left and right cant be the same index
            if left == -1 or left == right:
                continue
            
            seen = set()
            # iterate through the middle char 
            for i in range(left + 1, right):
                if s[i] not in seen:
                    count += 1
                    seen.add(s[i])
        return count 
    

        # GPT answer, ord will get the ASCII
        # returns the number correspoding the uniode code of a specified character
        # for c in map(chr, range(ord('a'), ord('z') + 1)):

        
sol = Solution()
tests = ["aabca", "adc", "bbcbaba"]
for s in tests:
    ans = sol.countPalindromicSubsequence(s)
    print(ans)

File: C++/Medium/1930.cpp
#include <string>
#include <iostream>
#include <unordered_set>

class Solution {
public:
    int countPalindromicSubsequence(std::string s) {
        int n = s.size();
        int count = 0;
        for (char c = 'a'; c <= 'z'; c++){
            int left = -1, right = -1;
            
            // find the first and the last occurence of c
            for(int i = 0; i < n; i++){
                if (s[i] == c){
                    if (left == -1){
                        left = i;
                    }
                    right = i;
                }
            }
            // left == -1 means didnt find c in s, left == right means there is only one occurence instead of 2
            if (left == -1 || left == right){
                continue;
            }

            bool seen[26] = {false};
            for(int i = left + 1; i < right; i++){
                char mid = s[i];
                // this is encoding mid to numbers, by - 'a', ex: 'a' - 'a' = 0, 'b' - 'a' = 1 
                int idx = mid - 'a';
                if (!seen[idx]){
                    count += 1;
                    seen[idx] = true;
                }
            }
        }
        return count;
    }
};